File: backend/app/api/github_repository_url.py
"""GitHub repository URL normalization at the intake/application boundary."""

from __future__ import annotations

from urllib.parse import urlsplit


class GitHubRepositoryUrlError(ValueError):
    def __init__(self, code: str, message: str) -> None:
        self.code = code
        super().__init__(message)


HTTPS_MESSAGE = "请输入 HTTPS GitHub 仓库地址"
HOST_MESSAGE = "目前仅支持 github.com 仓库"
CREDENTIALS_MESSAGE = "仓库地址不能包含用户名、密码或 Token"
PATH_MESSAGE = "请输入仓库主页地址，例如 https://github.com/owner/repo"


def normalize_github_repository_url(value: str) -> str:
    """Return a canonical https://github.com/owner/repo URL."""
    text = value.strip()
    if not text:
        raise GitHubRepositoryUrlError("https", HTTPS_MESSAGE)
    parsed = urlsplit(text)
    if parsed.scheme.lower() != "https":
        raise GitHubRepositoryUrlError("https", HTTPS_MESSAGE)
    if parsed.username or parsed.password:
        raise GitHubRepositoryUrlError("credentials", CREDENTIALS_MESSAGE)
    host = (parsed.hostname or "").rstrip(".").lower()
    if host == "www.github.com":
        host = "github.com"
    if host != "github.com":
        raise GitHubRepositoryUrlError("host", HOST_MESSAGE)
    if parsed.query or parsed.fragment:
        raise GitHubRepositoryUrlError("path", PATH_MESSAGE)
    segments = [item for item in parsed.path.split("/") if item]
    if len(segments) != 2:
        raise GitHubRepositoryUrlError("path", PATH_MESSAGE)
    owner, repo = segments
    if repo.lower().endswith(".git"):
        repo = repo[:-4]
    if not owner or not repo:
        raise GitHubRepositoryUrlError("path", PATH_MESSAGE)
    return f"https://github.com/{owner}/{repo}"

File: frontend/src/utils/githubRepositoryUrl.ts
export const GITHUB_REPOSITORY_URL_ERRORS = {
  https: "请输入 HTTPS GitHub 仓库地址",
  host: "目前仅支持 github.com 仓库",
  credentials: "仓库地址不能包含用户名、密码或 Token",
  path: "请输入仓库主页地址，例如 https://github.com/owner/repo",
} as const;

export type GitHubRepositoryUrlErrorCode = keyof typeof GITHUB_REPOSITORY_URL_ERRORS;

export class GitHubRepositoryUrlError extends Error {
  readonly code: GitHubRepositoryUrlErrorCode;

  constructor(code: GitHubRepositoryUrlErrorCode, message: string) {
    super(message);
    this.name = "GitHubRepositoryUrlError";
    this.code = code;
  }
}

export function normalizeGitHubRepositoryUrl(value: string): string {
  const text = value.trim();
  if (!text) {
    throw new GitHubRepositoryUrlError("https", GITHUB_REPOSITORY_URL_ERRORS.https);
  }

  let parsed: URL;
  try {
    parsed = new URL(text);
  } catch {
    throw new GitHubRepositoryUrlError("https", GITHUB_REPOSITORY_URL_ERRORS.https);
  }

  if (parsed.protocol !== "https:") {
    throw new GitHubRepositoryUrlError("https", GITHUB_REPOSITORY_URL_ERRORS.https);
  }
  if (parsed.username || parsed.password) {
    throw new GitHubRepositoryUrlError("credentials", GITHUB_REPOSITORY_URL_ERRORS.credentials);
  }

  let host = parsed.hostname.replace(/\.$/, "").toLowerCase();
  if (host === "www.github.com") {
    host = "github.com";
  }
  if (host !== "github.com") {
    throw new GitHubRepositoryUrlError("host", GITHUB_REPOSITORY_URL_ERRORS.host);
  }
  if (parsed.search || parsed.hash) {
    throw new GitHubRepositoryUrlError("path", GITHUB_REPOSITORY_URL_ERRORS.path);
  }

  const segments = parsed.pathname.split("/").filter(Boolean);
  if (segments.length !== 2) {
    throw new GitHubRepositoryUrlError("path", GITHUB_REPOSITORY_URL_ERRORS.path);
  }

  const owner = segments[0];
  let repo = segments[1];
  if (repo.toLowerCase().endsWith(".git")) {
    repo = repo.slice(0, -4);
  }
  if (!owner || !repo) {
    throw new GitHubRepositoryUrlError("path", GITHUB_REPOSITORY_URL_ERRORS.path);
  }
  return `https://github.com/${owner}/${repo}`;
}
